lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -0,0 +1,467 @@
+    let hash = show $ byronBlockHash b
+    putStrLn "\n\n"
+    putStrLn $ "=== SlotNo " ++ slot
+    putStrLn $ "=== " ++ hash
+    putStrLn "\n"
+    case byronBlockRaw b  of
+        CC.ABOBBlock cb -> do
+            let body = CC.blockBody cb
+            let txPay = CC.bodyTxPayload body
+            let txAuxs = CC.aUnTxPayload txPay
+            let txs = map CC.taTx txAuxs
+            print txs
+        CC.ABOBBoundary _ ->
+            putStrLn "Our friend the Epoch Boundary Block!"
+
+rollBackward :: Point ByronBlock -> IO ()
+rollBackward p = do
+    putStrLn $ "rollback to: " ++ show p
+
+
+-- A point where there is high activity on the mainnet
+point1 :: Point ByronBlock
+point1 =
+    let
+        Right h = CC.decodeAbstractHash "8a7505ec268a934c9c532fcf21185baf849c0de04a6d12cdf80b0fc45fe6ecb3"
+        slot = 21600*19+6540
+
+    in Abstract.Point $ Point.block
+        (Abstract.SlotNo slot)
+        (ByronHash h)
+
+-- How to run the node:
+--
+-- $ git clone https://github.com/input-output-hk/cardano-node
+-- $ cd cardano-node
+-- $ ./scripts/mainnet.sh
+--
+-- Then run main:
+main :: IO ()
+main = do
+    -- I have cardano-node running from ../cardano-node
+    let nodeId = CoreNodeId 0
+    let path = "../cardano-node/socket/" <> (localSocketFilePath nodeId)
+    let addr =  localSocketAddrInfo path
+
+    print point1
+
+    let params = ChainParameters
+         { epochSlots = EpochSlots 21600
+         , protocolMagic = ProtocolMagicId 764824073
+         }
+
+    let t = nullTracer
+    let client = OuroborosInitiatorApplication $ \pid -> \case
+            ChainSyncWithBlocksPtcl ->
+                chainSyncWithBlocks pid t params rollForward rollBackward
+            LocalTxSubmissionPtcl ->
+                (localTxSubmission pid t) >=> const (return ())
+    connectClient client dummyNodeToClientVersion addr
+
+--------------------------------------------------------------------------------
+--
+-- Concrete Types
+
+type Tx = GenTx ByronBlock
+
+data ChainParameters = ChainParameters
+    { epochSlots :: EpochSlots
+        -- ^ Number of slots per epoch.
+    , protocolMagic :: ProtocolMagicId
+        -- ^ Protocol magic (e.g. mainnet=764824073, testnet=1097911063)
+    }
+
+-- | Type representing a network client running two mini-protocols to sync
+-- from the chain and, submit transactions.
+type NetworkClient m = OuroborosApplication
+    'InitiatorApp
+        -- ^ Initiator ~ Client (as opposed to Responder / Server)
+    ConnectionId
+        -- ^ An identifier for the peer: here, a local and remote socket.
+    NodeToClientProtocols
+        -- ^ Specifies which mini-protocols our client is talking.
+        -- 'NodeToClientProtocols' allows for two mini-protocols:
+        --  - Chain Sync
+        --  - Tx submission
+    m
+        -- ^ Underlying monad we run in
+    ByteString
+        -- ^ Concrete representation for bytes string
+    ()
```

Comment:
I changed this from `Void` to `()` in hopes of being able to terminate the protocol with `SendMsgDone`. Not sure how effective it was.

I initially thought I could have `[ByronBlock]` here, and retrieve them in the end, but that did not seem like the case. It seems like `()` vs `Void` are the only useful types here.

Reply:
Seems like this return type is lost in `connectTo` which returns `IO ()`. https://github.com/input-output-hk/ouroboros-network/blob/9d8493e6d609f879c9ebc684bc5d830f1f146bb6/ouroboros-network/src/Ouroboros/Network/NodeToClient.hs#L161-L174

Reply:
Hmmm. The chain follower is never supposed to end, so `Void` should be the de-facto type here.